load-gen/src/main/java/io/smartcat/data/loader/LoadGenerator.java

```diff
@@ -1,155 +1,105 @@
-import io.smartcat.data.loader.tokenbuket.RefillStrategy;
-import io.smartcat.data.loader.tokenbuket.SleepStrategies;
-import io.smartcat.data.loader.tokenbuket.SleepStrategy;
-import io.smartcat.data.loader.tokenbuket.TokenBucket;
-import io.smartcat.data.loader.util.NoOpDataSource;
-import io.smartcat.data.loader.util.NoOpWorkTask;
+import io.smartcat.data.loader.api.RateGenerator;
+import io.smartcat.data.loader.api.Worker;
 
 /**
  * Load generator used to execute work tasks with data from provided data source.
+ *
+ * @param <T> Type of data which will be used.
  */
-public class LoadGenerator {
+public class LoadGenerator<T> {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(LoadGenerator.class);
+    private static final long NANOS_IN_SECOND = TimeUnit.SECONDS.toNanos(1);
+    private static final long TICK_PERIOD_IN_NANOS = 1000;
 
-    private final DataCollector dataCollector;
+    private final DataSource<T> dataSource;
+    private final RateGenerator rateGenerator;
+    private final Worker<T> worker;
 
-    private final TokenBucket tokenBucket;
-
-    private final PulseGenerator pulseGenerator;
-
-    private Timer metricsTimer;
-
-    private LoadGenerator(Builder builder) {
-        this.dataCollector = new DataCollector(builder.dataSource, builder.targetRate * 10);
-        this.tokenBucket = new TokenBucket(0, builder.refillStrategy, builder.sleepStrategy);
-        this.pulseGenerator = new PulseGenerator(dataCollector, tokenBucket, builder.workTask, builder.collectMetrics);
-
-        this.metricsTimer = new Timer("pulse-generator-timer");
-        this.metricsTimer.scheduleAtFixedRate(new MetricsLogger(), 0, 1000);
-    }
+    private AtomicBoolean terminate = new AtomicBoolean(false);
 
     /**
-     * Start load generator.
+     * Constructs load generator with specified <code>dataSource</code>, <code>rateGenerator</code> and
+     * <code>worker</code>.
+     *
+     * @param dataSource Data source from which load generator polls data.
+     * @param rateGenerator Rate generator which generates rate based on time point.
+     * @param worker Worker which accepts data polled from <code>dataSource</code> at rate provided by
+     *            <code>rateGenerator</code>.
      */
-    public void start() {
-        pulseGenerator.start();
+    public LoadGenerator(DataSource<T> dataSource, RateGenerator rateGenerator, Worker<T> worker) {
+        this.dataSource = dataSource;
+        this.rateGenerator = rateGenerator;
+        this.worker = worker;
     }
 
     /**
-     * Metrics logger timer task.
+     * Starts load generator.
+     *
+     * @throws IllegalStateException When start is attempted after load generator was already stopped.
      */
-    private class MetricsLogger extends TimerTask {
-        @Override
-        public void run() {
-            LOGGER.debug("Generated {} pulses", pulseGenerator.getPulseCount());
+    public void start() {
```

Comment:
I agree.  Should I leave stop() method named that way or rename it maybe to terminate?